inst/stan/functions/convolve.stan

```diff
@@ -1,48 +1,70 @@
-// convolve a pdf and case vector
-vector convolve(vector cases, vector rev_pmf) {
-    int t = num_elements(cases);
-    int max_pmf = num_elements(rev_pmf);
-    vector[t] conv_cases = rep_vector(1e-5, t);
-    for (s in 1:t) {
-        conv_cases[s] += dot_product(cases[max(1, (s - max_pmf + 1)):s],
-                                     tail(rev_pmf, min(max_pmf, s)));
+// convolve two vectors
+// length of x
+vector convolve(vector x, vector y, int len) {
```

Comment:
I think we have introduced a bug here as this convolves/adds two probability mass functions vs a backward rolling convolution of a vector and an associated PMF (as in the previous version). I think we likely need both. 

Trying to recreate synthetic data confirms there is likely a problem. 

![gp_nuts](https://user-images.githubusercontent.com/20317649/213675072-5223ec9d-85fe-4ad4-b0e9-bec21970404e.png)

I am just compiling main to check to see if this issue was actually introduced elsewhere (though due to the lack of commits touching the stan code I hope this isn't likely).

Reply:
i.e we have this: https://package.epinowcast.org/reference/add_pmfs.html (setting up some tests to check)

Reply:
Ah yes this looks like a bug and agree it's unlikely to be in `main`. Might be good to think about whether there's any test we can add (e.g. based on your example above) to catch instances where the model is being completely broken as seems to be the case here.

Reply:
yes thinking about it now. Just reading through model changes on `main` as well to check nothing small has been missed.

Reply:
So I looked at this on `main` and found that this behaviour was actually being driven by a bug in `simulate_infections()` which I have now corrected. Updating `develop` can reproduce synthetic data though it isn't entirely clear to me that it should be able to given the convolution implementation (perhaps it is equivalent and I am just not quite seeing it?). 

Regardless, I see that the synthetic validation takes around 2 hours on `develop` (https://github.com/epiforecasts/EpiNow2/actions/runs/3971839196) vs 35 minutes for `main` (https://github.com/epiforecasts/EpiNow2/actions/runs/3971832809) so it seems likely there is at the very least still a performance regression. 

![gp_nuts](https://user-images.githubusercontent.com/20317649/214002128-c709235f-086e-4119-8e7f-356359780927.png)

Reply:
So looking at the current vs old versions of `convolve` I think they are the same (just testing this explicitly now) and its just a matter of efficiency.

Reply:
Interesting - would you expect `dot_product` to be more efficient than the double loop?

Reply:
yes, I would with the sparse matrix version being the fastest based on experience with `epinowcast`/early versions of this code.

Reply:
I've added some tests to confirm things are working as expected. CI should show any performance difference.